File: python/python_asyncflow/generator.py
"""Base Generator classes for AsyncFlow components"""

import asyncio
import uuid
from abc import ABC, abstractmethod
from typing import Optional
from .logger import Logger


class Generator(ABC):
    """Base interface for all flow components"""
    
    @property
    @abstractmethod
    def id(self) -> str:
        """Unique identifier for this generator"""
        pass
    
    @property
    @abstractmethod
    def name(self) -> Optional[str]:
        """Optional name for debugging"""
        pass
    
    @name.setter
    @abstractmethod
    def name(self, name: str) -> None:
        """Set the name of this generator"""
        pass
    
    @abstractmethod
    def is_active(self) -> bool:
        """Check if generator is active"""
        pass
    
    @abstractmethod
    def is_running(self) -> bool:
        """Check if generator is running"""
        pass
    
    @abstractmethod
    def is_completed(self) -> bool:
        """Check if generator has completed"""
        pass
    
    @abstractmethod
    def activate(self) -> None:
        """Activate this generator"""
        pass
    
    @abstractmethod
    def deactivate(self) -> None:
        """Deactivate this generator"""
        pass
    
    @abstractmethod
    def complete(self) -> None:
        """Mark this generator as completed"""
        pass
    
    @abstractmethod
    async def step(self) -> None:
        """Execute one step of this generator"""
        pass
    
    @property
    @abstractmethod
    def logger(self) -> Logger:
        """Get the logger for this generator"""
        pass


class GeneratorBase:
    """Base implementation for Generator interface"""
    
    def __init__(self, name: Optional[str] = None):
        self._id = str(uuid.uuid4())
        self._name = name
        self._active = True
        self._running = True
        self._completed = False
        self._logger = Logger("AsyncFlow")
    
    @property
    def id(self) -> str:
        return self._id
    
    @property
    def name(self) -> Optional[str]:
        return self._name
    
    @name.setter
    def name(self, name: str) -> None:
        self._name = name
    
    def is_active(self) -> bool:
        return self._active
    
    def is_running(self) -> bool:
        return self._running
    
    def is_completed(self) -> bool:
        return self._completed
    
    def activate(self) -> None:
        self._active = True
    
    def deactivate(self) -> None:
        self._active = False
    
    def complete(self) -> None:
        self._completed = True
        self._running = False
    
    @property
    def logger(self) -> Logger:
        return self._logger

File: src/flow/generator.rs
use async_trait::async_trait;
use std::sync::atomic::{AtomicBool, Ordering};
use uuid::Uuid;
use crate::Logger;

#[async_trait]
pub trait Generator: Send + Sync {
    fn id(&self) -> Uuid;
    fn name(&self) -> Option<&str>;
    fn set_name(&mut self, name: String);
    fn is_active(&self) -> bool;
    fn is_running(&self) -> bool;
    fn is_completed(&self) -> bool;
    fn activate(&self);
    fn deactivate(&self);
    fn complete(&self);
    async fn step(&self) -> crate::Result<()>;
    fn logger(&self) -> &Logger;
}

pub struct GeneratorBase {
    id: Uuid,
    name: Option<String>,
    active: AtomicBool,
    running: AtomicBool,
    completed: AtomicBool,
    logger: Logger,
}

impl Clone for GeneratorBase {
    fn clone(&self) -> Self {
        Self {
            id: self.id,
            name: self.name.clone(),
            active: AtomicBool::new(self.active.load(Ordering::Relaxed)),
            running: AtomicBool::new(self.running.load(Ordering::Relaxed)),
            completed: AtomicBool::new(self.completed.load(Ordering::Relaxed)),
            logger: self.logger.clone(),
        }
    }
}

impl GeneratorBase {
    pub fn new() -> Self {
        Self {
            id: Uuid::new_v4(),
            name: None,
            active: AtomicBool::new(true),
            running: AtomicBool::new(true),
            completed: AtomicBool::new(false),
            logger: Logger::default(),
        }
    }

    pub fn with_name(name: impl Into<String>) -> Self {
        let mut base = Self::new();
        base.name = Some(name.into());
        base
    }

    pub fn id(&self) -> Uuid {
        self.id
    }

    pub fn name(&self) -> Option<&str> {
        self.name.as_deref()
    }

    pub fn set_name(&mut self, name: String) {
        self.name = Some(name);
    }

    pub fn is_active(&self) -> bool {
        self.active.load(Ordering::Relaxed)
    }

    pub fn is_running(&self) -> bool {
        self.running.load(Ordering::Relaxed)
    }

    pub fn is_completed(&self) -> bool {
        self.completed.load(Ordering::Relaxed)
    }

    pub fn activate(&self) {
        self.active.store(true, Ordering::Relaxed);
    }

    pub fn deactivate(&self) {
        self.active.store(false, Ordering::Relaxed);
    }

    pub fn complete(&self) {
        self.completed.store(true, Ordering::Relaxed);
        self.running.store(false, Ordering::Relaxed);
    }

    pub fn logger(&self) -> &Logger {
        &self.logger
    }
}